File: video/src/main/java/Dijkstra.java
import java.util.*;

public class Dijkstra {
    public static Map<String, Integer> calcularDistancias(Grafo grafo, String inicio, Map<String, String> previo) {
        Map<String, Integer> distancias = new HashMap<>();
        PriorityQueue<String> cola = new PriorityQueue<>(Comparator.comparingInt(distancias::get));

        for (String nodo : grafo.getAdyacencia().keySet()) {
            distancias.put(nodo, Integer.MAX_VALUE);
            previo.put(nodo, null);
        }
        distancias.put(inicio, 0);
        cola.add(inicio);

        while (!cola.isEmpty()) {
            String actual = cola.poll();
            for (Arista arista : grafo.getAdyacencia().get(actual)) {
                int nuevaDistancia = distancias.get(actual) + arista.peso;
                if (nuevaDistancia < distancias.get(arista.destino)) {
                    distancias.put(arista.destino, nuevaDistancia);
                    previo.put(arista.destino, actual);
                    cola.add(arista.destino);
                }
            }
        }
        return distancias;
    }

    public static List<String> reconstruirCamino(Map<String, String> previo, String inicio, String fin) {
        List<String> camino = new ArrayList<>();
        String actual = fin;
        while (actual != null) {
            camino.add(actual);
            actual = previo.get(actual);
        }
        Collections.reverse(camino);
        return camino;
    }
}

File: cali/dijkstra.py
"""
Implementación del algoritmo de Dijkstra usando listas de adyacencia.
Autor: Estudiante
"""

class Grafo:
    def __init__(self):
        # Diccionario para listas de adyacencia {nodo: [(vecino, peso), ...]}
        self.adyacencia = {}

    def agregar_nodo(self, nodo):
        if nodo not in self.adyacencia:
            self.adyacencia[nodo] = []

    def agregar_arista(self, origen, destino, peso):
        # Grafo no dirigido
        self.adyacencia[origen].append((destino, peso))
        self.adyacencia[destino].append((origen, peso))

    def dijkstra(self, inicio):
        import heapq
        distancias = {nodo: float('inf') for nodo in self.adyacencia}
        distancias[inicio] = 0
        previo = {nodo: None for nodo in self.adyacencia}
        cola = [(0, inicio)]

        while cola:
            distancia_actual, nodo_actual = heapq.heappop(cola)
            if distancia_actual > distancias[nodo_actual]:
                continue
            for vecino, peso in self.adyacencia[nodo_actual]:
                nueva_distancia = distancia_actual + peso
                if nueva_distancia < distancias[vecino]:
                    distancias[vecino] = nueva_distancia
                    previo[vecino] = nodo_actual
                    heapq.heappush(cola, (nueva_distancia, vecino))
        return distancias, previo

    def reconstruir_camino(self, previo, inicio, fin):
        camino = []
        actual = fin
        while actual is not None:
            camino.append(actual)
            actual = previo[actual]
        return list(reversed(camino))
